### PR TITLE
Remove player from game and queue if they are moved to spectator

### PR DESCRIPTION
**Pull request by dollannn**
I made these changes to ensure it is compatible with a plugin such as [AFKManager](https://github.com/NiGHT757/AFKManager) which only moves a player to spectator and has no way of knowing how to dequeue a player. Without this change players will be moved to spectator and then moved back into a team and actually never get kicked if you do not use an overly aggressive afk kick of 15 seconds without spectator. I made this for personal use but thought it might be benefit others

### CODE REVIEW

**Comment by B3none** [comment]
These changes look great, are you able to move this logic into the relevant manager, I think I have separate team management logic in one of those where this code should be. If it can’t go there and it must go here then we can, I would also like to add a config option to give people the choice for how they would like to handle this. Basically just put it in an if statement with a config variable check.

**Comment by dollannn** [comment]
Hope this resolves your comment, was unsure where to place the function but currently I put it in the GameManager class as it best fits there.

File: README.md [hunk 1]
```diff
@@ -2,14 +2,17 @@
 ![GitHub Actions Workflow Status](https://img.shields.io/github/actions/workflow/status/b3none/cs2-retakes/plugin-build.yml?branch=master&style=flat-square&label=Latest%20Build)
 
 # CS2 Retakes
+
```
Comment (B3none) [inline README.md:5]: Revert this additional line, and others formatting changes to the README

File: RetakesPlugin/Modules/Managers/GameManager.cs [hunk 2]
```diff
@@ -11,17 +12,18 @@ public class GameManager
     public readonly QueueManager QueueManager;
     private readonly int _consecutiveRoundWinsToScramble;
     private readonly bool _isScrambleEnabled;
-
+    private readonly bool _RemoveSpectatorsEnabled;
```
Comment (B3none) [inline RetakesPlugin/Modules/Managers/GameManager.cs:15]: Rename this variable to `_removeSpectatorsEnabled`

File: RetakesPlugin/Modules/Managers/GameManager.cs [hunk 3]
```diff
@@ -1,6 +1,7 @@
 ﻿using CounterStrikeSharp.API;
 using CounterStrikeSharp.API.Core;
 using CounterStrikeSharp.API.Modules.Utils;
+using Serilog.Debugging;
```
Comment (B3none) [inline RetakesPlugin/Modules/Managers/GameManager.cs:4]: Remove this, and it's usages

File: RetakesPlugin/Modules/Configs/RetakesConfigData.cs [hunk 4]
```diff
@@ -19,4 +19,6 @@ public class RetakesConfigData
     public bool IsDebugMode { get; set; } = false;
     public bool ShouldForceEvenTeamsWhenPlayerCountIsMultipleOf10 { get; set; } = true;
     public bool EnableFallbackBombsiteAnnouncement { get; set; } = true;
+    public bool RemoveSpectators { get; set; } = true;
+
```
Comment (B3none) [inline RetakesPlugin/Modules/Configs/RetakesConfigData.cs:23]: Remove the additional extra line

File: RetakesPlugin/Modules/Managers/GameManager.cs [hunk 5]
```diff
@@ -11,17 +12,18 @@ public class GameManager
...
     public const int ScoreForDefuse = 50;
 
-    public GameManager(Translator translator, QueueManager queueManager, int? roundsToScramble, bool? isScrambleEnabled)
+    public GameManager(Translator translator, QueueManager queueManager, int? roundsToScramble, bool? isScrambleEnabled, bool? RemoveSpectatorsEnabled)
     {
         _translator = translator;
         QueueManager = queueManager;
         _consecutiveRoundWinsToScramble = roundsToScramble ?? 5;
         _isScrambleEnabled = isScrambleEnabled ?? true;
+        _RemoveSpectatorsEnabled = RemoveSpectatorsEnabled ?? true;
```
Comment (B3none) [inline RetakesPlugin/Modules/Managers/GameManager.cs:26]: You need to default this behaviour to false, as that's how it was before and I want to maintain backwards compatibility / not introduce a potentially breaking change to peoples servers.

File: README.md [hunk 6]
```diff
@@ -64,8 +73,10 @@ When the plugin is first loaded it will create a `retakes_config.json` file in t
 | IsDebugMode                                       | Whether to enable debug output to the server console or not.                                                                                    | false      | false      | true       |
 | ShouldForceEvenTeamsWhenPlayerCountIsMultipleOf10 | Whether to force even teams when the active players is a multiple of 10 or not. (this means you will get 5v5 @ 10 players / 10v10 @ 20 players) | true       | false      | true       |
 | EnableFallbackBombsiteAnnouncement                | Whether to enable the fallback bombsite announcement.                                                                                           | true       | false      | true       |
+| RemoveSpectators                                 | When a player is moved to spectators, remove them from all retake queues. Ensures that AFK plugins work as expected.                                                         | true       | false      | true       |
```
Comment (B3none) [inline README.md:76]: Default the value to false

File: RetakesPlugin/Modules/Configs/RetakesConfigData.cs [hunk 7]
```diff
@@ -19,4 +19,6 @@ public class RetakesConfigData
     public bool IsDebugMode { get; set; } = false;
     public bool ShouldForceEvenTeamsWhenPlayerCountIsMultipleOf10 { get; set; } = true;
     public bool EnableFallbackBombsiteAnnouncement { get; set; } = true;
+    public bool RemoveSpectators { get; set; } = true;
```
Comment (B3none) [inline RetakesPlugin/Modules/Configs/RetakesConfigData.cs:22]: Default this value to false, also as it's a boolean it needs to be prefixed with `Is` / `Has` / `Should` / `Enable`

File: RetakesPlugin/RetakesPlugin.cs [hunk 8]
```diff
@@ -815,7 +816,12 @@ public HookResult OnPlayerTeam(EventPlayerTeam @event, GameEventInfo info)
         // Ensure all team join events are silent.
         @event.Silent = true;
 
-        return HookResult.Continue;
+        if (_gameManager == null)
+        {
+            Helpers.Debug($"Game manager not loaded.");
+            return HookResult.Continue;
+        }
+        return _gameManager.RemoveSpectators(@event, _hasMutedVoices);
```
Comment (B3none) [inline RetakesPlugin/RetakesPlugin.cs:824]: Add an empty line above this return statement

**Review by B3none — CHANGES_REQUESTED** [review]
Some minor tweaks, once those are addressed I'll make a release. Also please bump the plugin version const to `2.0.3` in `RetakesPlugin.cs`

**Comment by dollannn** [comment]
Bumped to 2.0.5 instead of 2.0.3 as we are on 2.0.4

File: README.md [hunk 9]
```diff
@@ -64,6 +64,7 @@ When the plugin is first loaded it will create a `retakes_config.json` file in t
 | IsDebugMode                                       | Whether to enable debug output to the server console or not.                                                                                    | false      | false      | true       |
 | ShouldForceEvenTeamsWhenPlayerCountIsMultipleOf10 | Whether to force even teams when the active players is a multiple of 10 or not. (this means you will get 5v5 @ 10 players / 10v10 @ 20 players) | true       | false      | true       |
 | EnableFallbackBombsiteAnnouncement                | Whether to enable the fallback bombsite announcement.                                                                                           | true       | false      | true       |
+| RemoveSpectators                                  | When a player is moved to spectators, remove them from all retake queues. Ensures that AFK plugins work as expected.                            | false      |    false      | true       |
```
Comment (B3none) [inline README.md:67]: Update the name of the variable in the readme

File: README.md [hunk 10]
```diff
@@ -83,4 +84,4 @@ Subscribe to **release** notifications and stay up to date with the latest featu
 ![image](https://github.com/B3none/cs2-retakes/assets/24966460/e288a882-0f1f-4e8c-b67f-e4c066af34ea)
 
 ## Credits
-This was inspired by the [CS:GO Retakes project](https://github.com/splewis/csgo-retakes) written by [splewis](https://github.com/splewis).
+This was inspired by the [CS:GO Retakes project](https://github.com/splewis/csgo-retakes) written by [splewis](https://github.com/splewis).
```
Comment (B3none) [inline README.md:87]: Remove this formatting change

**Review by B3none — CHANGES_REQUESTED** [review]
Well spotted with the version, two minor tweaks and it’s good to merge.

**Review by B3none — APPROVED** [review]

**Pull request merged**